### PR TITLE
Feature child device operations

### PR DESCRIPTION
**Pull request by sw-tt-tusharojha**

### CODE REVIEW

File: IOTConnectSDK/app/src/main/java/com/softweb/iotconnectsdk/activity/GatewayChildDevicesActivity.kt [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
+
+    }
+
+    override fun onNothingSelected(parent: AdapterView<*>?) {
+
+    }
+
+    override fun onDeviceCommand(message: String?) {
+        TODO("Not yet implemented")
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/app/src/main/java/com/softweb/iotconnectsdk/activity/GatewayChildDevicesActivity.kt]: @sw-tt-rajkiranvank @sw-tt-tusharojha this functionality is not in scope?

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/SDKClient.kt [hunk 2]
```diff
@@ -202,5 +228,49 @@ class SDKClient(
         sdkClientManager?.sendData(jsonData)
     }
 
+
+    /**
+     *Get child device from IOT connect portal
+     *https://docs.iotconnect.io/iotconnect/resources/device-message-2-1-2/device-identity-messages/#devices
+     *
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/SDKClient.kt:234]: @sw-tt-tusharojha @sw-tt-rajkiranvank  add more documentation

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/enums/C2DMessageEnums.kt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 package com.iotconnectsdk.enums
 
-enum class C2DMessageEnums(val value: Int) {
+internal enum class C2DMessageEnums(val value: Int) {
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/enums/C2DMessageEnums.kt:3]: @sw-tt-rajkiranvank @sw-tt-tusharojha what will be default code if you don't receive from your list?

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/mqtt/IotSDKMQTTService.kt [hunk 4]
```diff
@@ -129,12 +151,84 @@ internal class IotSDKMQTTService private constructor(
...
-        if (protocolBean.pwd != null) {
-            mqttConnectOptions.password = protocolBean.pwd.toCharArray()
+
+        var sdkObj: JSONObject
+        try {
+            if (sdkOptions != null) {
+                sdkObj = JSONObject(sdkOptions)
+                if (authenticationType == IotSDKUrls.AUTH_TYPE_SELF_SIGN || authenticationType == IotSDKUrls.AUTH_TYPE_CA_SIGN) {
+                    if (sdkObj.has("certificate")) {
+                        val certificate = sdkObj.getJSONObject("certificate")
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/mqtt/IotSDKMQTTService.kt:203]: @sw-tt-rajkiranvank @sw-tt-tusharojha  what happens if certificate name is different

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt [hunk 5]
```diff
@@ -179,4 +221,60 @@ internal object SDKClientUtils {
         }
         return true
     }
+
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    @Throws(Exception::class)
+    fun generateSasTokenLatest(resourceUri: String?, key: String?): String? {
+        // Token will expire in one year
+        val expiry = Instant.now().epochSecond + 31536000
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt]: @sw-tt-rajkiranvank @sw-tt-tusharojha take this value in constant 31536000

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt [hunk 6]
```diff
@@ -179,4 +221,60 @@ internal object SDKClientUtils {
...
+                + "&sig=" + URLEncoder.encode(
+            signature,
+            StandardCharsets.UTF_8.name()
+        ) + "&se=" + expiry)
+    }
+
+    @Throws(java.lang.Exception::class)
+    fun generateSasToken(resourceUri: String?, key: String?): String? {
+        // Token will expire in one year
+        val expiry = Instant.now().epochSecond + 31536000
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt]: @sw-tt-rajkiranvank @sw-tt-tusharojha same here

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SecurityHelper.kt [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        // In this case, we are working without a client key password
+        return try {
+            Security.addProvider(BouncyCastleProvider())
+            val keyManagers = createKeyManagerFactory(
+                clientCrtFile,
+                clientKeyFile,
+                clientKeyPassword,
+                companySID,
+                deviceUniqueID
+            )!!.keyManagers
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SecurityHelper.kt]: @sw-tt-tusharojha @sw-tt-rajkiranvank !! will create a 'NullPointerException'  can you please write in better way

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/ValidationUtils.kt [hunk 8]
```diff
@@ -177,11 +177,11 @@ internal class ValidationUtils private constructor(
         return true
     }
 
-    fun validateAckParameters(obj: JSONObject?, messageType: String): Boolean {
-        if (obj == null || messageType.isEmpty()) {
+    fun validateAckParameters(obj: JSONObject?): Boolean {
+       /* if (obj == null || messageType.isEmpty()) {
```
Comment (sw-ot-deveshmevada) [inline IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/ValidationUtils.kt]: @sw-tt-tusharojha @sw-tt-rajkiranvank  remove unwanted code

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt (hunk 5 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt]: @sw-ot-deveshmevada 
Issue resolved 
Value taken in constant

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt (hunk 6 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SDKClientUtils.kt]: @sw-ot-deveshmevada
Issue resolved
Value taken in constant

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/ValidationUtils.kt (hunk 8 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/ValidationUtils.kt]: @sw-ot-deveshmevada
Issue resolved
Unwanted code removed

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/SDKClient.kt (hunk 2 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/SDKClient.kt:234]: @sw-ot-deveshmevada 
Comment added for the function

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SecurityHelper.kt (hunk 7 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/utils/SecurityHelper.kt]: @sw-ot-deveshmevada 
Issue resolved **!!**  change to this **?.** now NullPointerException will not be created

File: IOTConnectSDK/app/src/main/java/com/softweb/iotconnectsdk/activity/GatewayChildDevicesActivity.kt (hunk 1 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/app/src/main/java/com/softweb/iotconnectsdk/activity/GatewayChildDevicesActivity.kt]: HI @sw-ot-deveshmevada 
I have put comment so user can understand that

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/enums/C2DMessageEnums.kt (hunk 3 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/enums/C2DMessageEnums.kt:3]: Hi @sw-ot-deveshmevada 
I have put the else condition accordingly

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/mqtt/IotSDKMQTTService.kt (hunk 4 above)
Comment (sw-tt-tusharojha) [reply IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/mqtt/IotSDKMQTTService.kt:203]: Hi @sw-ot-deveshmevada 
I have made sdk options and environment variable accessible from sdk only. Now he is not able to change keys

**Review by sw-ot-deveshmevada — APPROVED** [review]

**Pull request merged**